File: app/src/main/java/com/theandroiddev/riffking/ThreadAdapter.java
package com.theandroiddev.riffking;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.mikhaellopez.circularimageview.CircularImageView;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

/**
 * Created by jakub on 18.07.17.
 */

public class ThreadAdapter extends RecyclerView.Adapter<ThreadAdapter.ViewHolder>
        implements AsyncResponse {
    private static final String TAG = "ThreadAdapter";
    DatabaseReference mDatabase;
    private ArrayList<Thread> threads;
    private Context context;


    public ThreadAdapter(Context context, ArrayList<Thread> threads, DatabaseReference mDatabase) {
        this.context = context;
        this.threads = threads;
        this.mDatabase = mDatabase;

    }


    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.single_thread,
                parent, false);

        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        Log.d(TAG, "Element " + position + " set.");

        holder.getThreadTitleTv().setText(threads.get(position).getTitle());
        setThreadCreator((position), holder.threadUserTv, holder.threadUserIv);

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //Toast.makeText(context, String.valueOf(position), Toast.LENGTH_SHORT).show();
                openThread(holder.getAdapterPosition());
            }
        });

        holder.threadStatsTv.setText(initStats(threads.get(position)));
        holder.threadDateTv.setText(initDate(threads.get(position)));

        new GetThumbnail(threads.get(position).getVideoUrl(), holder.threadThumbnailTv).execute();

    }

    private String initStats(Thread thread) {

        return thread.getViews() + " views " + thread.getLikes() + " likes";

    }

    private String initDate(Thread thread) {

        return thread.getDate();

    }


    private void openThread(int position) {

        Bundle bundle = new Bundle();
        bundle.putString("threadId", String.valueOf(threads.get(position).getId()));

        android.support.v4.app.Fragment threadFragment = new ThreadFragment();
        threadFragment.setArguments(bundle);
        try {
            FragmentManager fragmentManager = ((FragmentActivity) context).getSupportFragmentManager();
            FragmentTransaction transaction = fragmentManager.beginTransaction();
            transaction.replace(R.id.content_home, threadFragment).addToBackStack("stack1");
            transaction.commit();

        } catch (ClassCastException e) {
            Log.e(TAG, "cant get fragment manager");
        }
    }


    @Override
    public int getItemCount() {
        return threads.size();
    }

    @Override
    public void processFinish(Bitmap output) {


    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    public void setThreadCreator(final int position, final TextView tv, final CircularImageView iv) {

        mDatabase.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {

                tv.setText(dataSnapshot.child("users").child(threads.get(position).getUserId()).child("name").getValue(String.class));
                Picasso.with(context).load(dataSnapshot.child("users").child(threads.get(position).getUserId()).child("photoUrl").getValue(String.class))
                        .into(iv);
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });

    }


    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView threadTitleTv, threadStatsTv, threadUserTv;
        private final TextView threadDateTv;
        private final ImageView threadThumbnailTv;
        private final CircularImageView threadUserIv;

        public ViewHolder(View itemView) {
            super(itemView);

            threadTitleTv = (TextView) itemView.findViewById(R.id.text_view);
            threadUserTv = (TextView) itemView.findViewById(R.id.nick_tv);
            threadThumbnailTv = (ImageView) itemView.findViewById(R.id.thumbnail_img);
            threadStatsTv = (TextView) itemView.findViewById(R.id.stats_tv);
            threadDateTv = (TextView) itemView.findViewById(R.id.date_tv);
            threadUserIv = (CircularImageView) itemView.findViewById(R.id.nick_iv);

        }

        public TextView getThreadTitleTv() {
            return threadTitleTv;
        }
    }


}

File: app/src/main/java/com/theandroiddev/riffking/ThreadAdapter.kt
package com.theandroiddev.riffking

import android.content.Context
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import com.mikhaellopez.circularimageview.CircularImageView
import com.squareup.picasso.Picasso
import java.util.*

/**
 * Created by jakub on 18.07.17.
 */

class ThreadAdapter(private val context: Context, private val threads: ArrayList<Thread>, internal var mDatabase: DatabaseReference) : RecyclerView.Adapter<ThreadAdapter.ViewHolder>(), AsyncResponse {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        val v = LayoutInflater.from(parent.context).inflate(R.layout.single_thread,
                parent, false)

        //TODO set ThumbnailIv size on start

        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        Log.d(TAG, "Element $position set.")

        val videoUrl = threads[position].videoUrl
        if (videoUrl != null) {
            GetThumbnail(videoUrl, holder.threadThumbnailIv).execute()
        }
        holder.threadTitleTv.text = threads[position].title
        setThreadCreator(position, holder.threadUserTv, holder.threadUserIv)

        holder.itemView.setOnClickListener {
            //Toast.makeText(context, String.valueOf(position), Toast.LENGTH_SHORT).show();
            openThread(holder.adapterPosition)
        }
        holder.threadStatsTv.text = initStats(threads[position])
        holder.threadDateTv.text = initDate(threads[position])

    }

    private fun initStats(thread: Thread): String {

        return thread.views.toString() + " views " + thread.likes + " likes"

    }

    private fun initDate(thread: Thread): String? {

        return thread.date

    }

    private fun openThread(position: Int) {

        val bundle = Bundle()
        bundle.putString("threadId", threads[position].id.toString())

        val threadFragment = ThreadFragment()
        threadFragment.arguments = bundle
        try {
            val fragmentManager = (context as FragmentActivity).supportFragmentManager
            val transaction = fragmentManager.beginTransaction()
            transaction.replace(R.id.content_home, threadFragment).addToBackStack("stack1")
            transaction.commit()

        } catch (e: ClassCastException) {
            Log.e(TAG, "cant get fragment manager")
        }

    }


    override fun getItemCount(): Int {
        return threads.size
    }

    override fun processFinish(output: Bitmap) {

    }

    fun setThreadCreator(position: Int, tv: TextView, iv: CircularImageView) {

        mDatabase.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {

                //Picasso.with(iv.getContext()).cancelRequest(iv);

                tv.text = dataSnapshot.child("users").child(threads[position].userId!!).child("name").getValue(String::class.java)
                //                Picasso.with(context).load(dataSnapshot.child("users").child(threads.get(position).getUserId()).child("photoUrl").getValue(String.class))
                //                        .into(iv);
                Picasso.get().load(dataSnapshot.child("users").child(threads[position]
                        .userId!!).child("photoUrl").getValue(String::class.java)).into(iv)

            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })

    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val threadTitleTv: TextView
        val threadStatsTv: TextView
        val threadUserTv: TextView
        val threadDateTv: TextView
        val threadThumbnailIv: ImageView
        val threadUserIv: CircularImageView
        val threadLayout: LinearLayout

        init {

            threadTitleTv = itemView.findViewById<View>(R.id.text_view) as TextView
            threadUserTv = itemView.findViewById<View>(R.id.nick_tv) as TextView
            threadStatsTv = itemView.findViewById<View>(R.id.stats_tv) as TextView
            threadDateTv = itemView.findViewById<View>(R.id.date_tv) as TextView
            threadUserIv = itemView.findViewById<View>(R.id.nick_iv) as CircularImageView
            threadThumbnailIv = itemView.findViewById<View>(R.id.thumbnail_img) as ImageView
            threadLayout = itemView.findViewById<View>(R.id.thread_layout) as LinearLayout
        }
    }

    companion object {
        private val TAG = "ThreadAdapter"
    }

}
